File: calc_dup/make_run.pl
#!/bio/bin/perl -w
use File::Basename;

my $RUN=378; #
my $j=1;
my $NO="_1";
my @SAMPLE_LIST=(7308..7339); #
my $PREFIX="Project_";
my $SAMPLE_NAME="";
my @RUN_NO=("PE001");
my @RUN_DIR=(
"/data/HiSeq2000/160216_D00734_0060_BC8NAEACXX/Unaligned/"
);

##my $SURE_POS="/grid2/personal-genome/BED/SureSelect_All_Exon_50mb_with_annotation_hg19_bed";
#my $SUFFIX="_TruSeq";
#my $SURE_POS="/grid2/personal-genome/BED/TruSeq-Exome-Targeted-Regions-BED-file";
#my $SUFFIX="_SS4UTR";
#my $SURE_POS="/grid2/personal-genome/BED/SureSelect_All_Exon_V4+UTRs_hg19.bed";
#my $SUFFIX="_SS5UTR";
#my $SURE_POS="/grid2/personal-genome/BED/S04380219_V5_Core+UTR_Fragments.bed";
my $SUFFIX="_SS6UTR";
my $SURE_POS="/grid2/personal-genome/BED/S07604624_V6+UTR_Covered.bed";
#my $SUFFIX="_Amplicon";
#my $SURE_POS="";
#my $SUFFIX="_WG";
#my $SURE_POS="";
#my $SUFFIX="_Custom";
#my $SURE_POS="0494601_Covered-205_custom.bed";
#my $SUFFIX="_RNA";
#my $SURE_POS="";
#my $SUFFIX="_Cancer";
#my $SURE_POS="/grid2/personal-genome/BED/120522_HG19_Onco_R_EZ.bed";

my $RNA_FLAG = 0;
if($SUFFIX eq "_RNA"){
    $RNA_FLAG = 1;
}

foreach my $SAMPLE (@SAMPLE_LIST){ # START of each sample loop
    
    print "SAMPLE: $RUN/$SAMPLE$NO$SUFFIXn";
    $SAMPLE_NAME .= " $SAMPLE$NO$SUFFIX";
    
    my $GENOME_FLAG = 0;
    my $GENOME = "exome";
    if($SUFFIX eq "_WG"){
        $GENOME_FLAG = 1;
        $GENOME = "genome";
    }
    if($RNA_FLAG){
        $GENOME = "rna";
    }
    
    my @RUN_PE=();
    my @RUN_DIR_PE=();
    my @RUN_SE=();
    my @RUN_DIR_SE=();
    
    for(my $i=0;$i<=$#RUN_DIR;$i++){
        my $DIR = $RUN_DIR[$i];
        my @READ_DIR = glob($DIR . "/" . $PREFIX . $SAMPLE . "*" . $SUFFIX . "/Sample_" . $SAMPLE . "*");
        
        foreach(@READ_DIR){
            chomp;
            if(!-d $_){
                print "Error: $_ was not found!n";
                next;
            }
            if($RUN_NO[$i] =~ /^PE/){
                push(@RUN_PE, $RUN_NO[$i]);
                push(@RUN_DIR_PE, $_);
            }
            else{
                push(@RUN_SE, $RUN_NO[$i]);
                push(@RUN_DIR_SE, $_);
            }
        }
    }
    
    $SAMPLE .= $NO . $SUFFIX;
    
    my $FASTQ_DIR="$RUN/$SAMPLE/$GENOME/fastq";
    !system("mkdir -p $FASTQ_DIR") || die "Error mkdirn";
    
    my $FASTQ_PE="";
    my $FASTQ_SE="";
    #&check();
    
    for(my $i=0; $i<=$#RUN_PE; $i++){
        while(glob("$RUN_DIR_PE[$i]/*.fastq.gz")){
            my @temp=split("_", basename($_));
            
            $temp[4]=~s/R//g;
            $temp[5]=~s/.fastq.gz//g;
            
            $FILE_BASE="$SAMPLE.$RUN_PE[$i].$temp[3]_$temp[5]_$temp[4]";
            $cmd="ln -s $_ $FASTQ_DIR/$FILE_BASE.fastq.gz";
            #    print "$cmdn";
            !system("$cmd") || die "Errorn";
            
            if($temp[4] eq "1"){
                $FILE_BASE="$SAMPLE.$RUN_PE[$i].$temp[3]_$temp[5]";
                
                if($FASTQ_PE eq ""){
                    $FASTQ_PE="$FILE_BASE";
                    }else{
                    $FASTQ_PE="$FASTQ_PE $FILE_BASE";
                }
            }
        }
    }
    
    for(my $i=0; $i<=$#RUN_SE; $i++){
        while(glob("$RUN_DIR_SE[$i]/*_R1_*.fastq.gz")){
            my @temp=split("_", basename($_));
            
            $temp[4]=~s/R//g;
            $temp[5]=~s/.fastq.gz//g;
            
            $FILE_BASE="$SAMPLE.$RUN_SE[$i].$temp[3]_$temp[5]_$temp[4]";
            $cmd="ln -s $_ $FASTQ_DIR/$FILE_BASE.fastq.gz";
            print "$cmdn";
            !system("$cmd") || die "Errorn";
            
            if($temp[4] eq "1"){
                $FILE_BASE="$SAMPLE.$RUN_SE[$i].$temp[3]_$temp[5]";
                
                if($FASTQ_SE eq ""){
                    $FASTQ_SE="$FILE_BASE";
                    }else{
                    $FASTQ_SE="$FASTQ_SE $FILE_BASE";
                }
            }
        }
    }
    
    ### make run.sh
    if(!$RNA_FLAG){
        my $thread=`gxpc e hostname | wc -l`;
        $thread=~s/n//g;
        
        my $RUN_FILE="$RUN/$SAMPLE/run.sh";
        open(OUT,">$RUN_FILE") || die "Errorn";
        
        print OUT "gxpc make -k -j $thread -f /nfs/6/personal-genome/makefiles/makefile_md5.nfs ";
        print OUT "SAMPLE=$SAMPLE PE_READS="$FASTQ_PE" SE_READS="$FASTQ_SE" ";
        print OUT "SURE_POS=$SURE_POS";
        print OUT " TARGET=genome" if($GENOME_FLAG);
        print OUT "n";
        close(OUT);
    }
    
} # END of each sample loop

###  make run script
my $RUN_SCRIPT="$RUN/auto_run$SUFFIX.sh";
open(OUT,">$RUN_SCRIPT") || die "Cannot create $RUN_SCRIPT";
print OUT "BASE_DIR=/work/HiSeq2000/$RUN
for DIR in $SAMPLE_NAME
do
echo $DIRn";

if(!$RNA_FLAG){
    print OUT "tcd $BASE_DIR/$DIR && date >> make.log && (time sh run.sh) >> make.log 2>> make.logn";
    print OUT "tcd $BASE_DIR/ && sh /work/yoshimura/tools/check_results.sh $DIR >> check_results.log 2>> check_results.logn";
}
else{
    print OUT "tqsub -N RNA-$DIR -o $DIR/make.log -j y /work/HiSeq2000/BaseCall/qsub_rna.sh $DIRn";
}

print OUT "done";
close(OUT);
#sub check{
    #        die("Check SAMPLE!n") if($SAMPLE eq "0242_1_SS4UTR");
    #        for(my $i=0; $i<=$#RUN_PE; $i++){
        #                die("$RUN_DIR_PE[$i] does not exist!n") if(! -d $RUN_DIR_PE[$i]);
        #                die("check RUN_DIR_PE[$i]!n") if($RUN_DIR_PE[$i] !~ /$SAMPLE/);
    #        }
    #        for(my $i=0; $i<=$#RUN_SE; $i++){
        #               die("$RUN_DIR_SE[$i] does not exist!n") if(! -d $RUN_DIR_SE[$i]);
        #               die("check RUN_DIR_SE[$i]!n") if($RUN_DIR_SE[$i] !~ /$SAMPLE/);
    #       }
#}

File: calc_dup/make_run.rb
#!/bio/bin/ruby

require 'yaml'
require 'optparse'

require_relative '../app/prepkit.rb'
require_relative '../app/config.rb'
PREP = Prepkit.new( load_config().prepkit_info )

############### get args, todo: prefer _ than - in log opt

if __FILE__ == $0 

OPT = {}
opt = OptionParser.new
opt.on( '--library-ids V' , desc = "',' separeted id list" ){|v| OPT[:libids_raw] = v } # ex. 7381,7309_2
opt.on( '--run V', String ){|v| OPT[:run] = v } # slide ex. 256
opt.on( '--run-name V', String ){|v| OPT[:run_name] = v } # ex. 372813_sb38310_dsa23131sda
opt.on( '--suffix V' ){|v| OPT[:suffix] = v } # ex. SS5UTR
opt.on( '--storage V' ){|v| OPT[:storage] = v } # path
opt.on( '--path-check-result V' ){|v| OPT[:path_check] = v } # needed?
opt.on( '--path-makefile V' ){|v| OPT[:path_makefile] = v } # path

opt.parse!(ARGV)
for req in [:libids_raw, :run, :run_name, :suffix, :storage, :path_check, :path_makefile]
  unless OPT[req]
    raise "missing option; #{req}"
  end
end

run = OPT[:run]
run_name = OPT[:run_name]
suffix = OPT[:suffix]
libid_list = OPT[:libids_raw].split(/,/)
path_makefile = OPT[:path_makefile]
storage = OPT[:storage]
path_check = OPT[:path_check]

make_run_sh(run, run_name, suffix, libid_list, storage, path_makefile, path_check)
end

def make_run_sh(run, run_name, suffix, libid_list, storage, path_makefile, path_check)

run_dir = [ "/data/HiSeq2000/#{run_name}/Unaligned/" ]
prefix = 'Project_'
sample_names = ''
run_no=['PE001']
sure_pos = PREP.suffix2sure_pos( suffix )
rna_flag = (suffix == '_RNA')? true : false


###############

for sample in libid_list
  
  puts "SAMPLE: #{run}/#{sample}#{suffix}"
  sample_names += " #{sample}#{suffix}"
  
  genome_flag , genome = false, 'exome'
  genome_flag , genome = true, 'genome' if suffix == '_WG'
  genome = 'rna' if rna_flag

  run_pe , run_dir_pe = [] , []
  run_se , run_dir_se = [] , []
  
  ############### loop to push elements to RUN[_DIR]_[PS]E
  run_dir.each_with_index do |dir, i|
    for direc in Dir.glob( File.join( dir, "#{prefix}#{sample}*#{suffix}", "Sample_#{sample}*" ) )
      direc.chomp!
      unless Dir.exists? direc
        puts "Error; #{direc} was not found"
        next
      end
      if run_no[i].match /^PE/
        run_pe << run_no[i]
        run_dir_pe << direc
      else
        run_se << run_no[i]
        run_dir_se << direc
      end
    end
  end
  
  ### !!! ###
  sample += suffix

  fastq_dir = File.join( storage, run, sample, genome, 'fastq' )
  # `` system call -- exit if fail 
  # system() -- exit ruby proess if fail
  `mkdir -p #{fastq_dir}`
  
  fastq_pe = []
  ############### run_pe loop to make fastq_pe
  for pe, pe_dir in run_pe.zip(run_dir_pe)
    for fst in Dir.glob( File.join( pe_dir, "*.fastq.gz" ) )
      temp = File.basename( fst ).split("_")
      temp[4].gsub! /R/, ''
      temp[5].gsub! /.fastq.gz/, ''
      file_base = "#{sample}.#{pe}.#{temp[3]}_#{temp[5]}_#{temp[4]}"

      cmd = "ln -s #{fst} #{fastq_dir}/#{file_base}.fastq.gz"
      # STDERR.puts cmd
      `#{cmd}`

      if temp[4] == '1'
        fastq_pe << "#{sample}.#{pe}.#{temp[3]}_#{temp[5]}"
      end

    end
  end
  puts " --- #{fastq_pe}"
  fastq_pe = fastq_pe.join(" ")

  fastq_se = []
  ############### run_se loop to make fastq_se
  for se, se_dir in run_se.zip(run_dir_se)
    for fst in Dir.glob( File.join( se_dir, "*_R1_*.fastq.gz" ) )
      temp = File.basename( fst ).split("_")
      temp[4].gsub! /R/, ''
      temp[5].gsub! /.fastq.gz/, ''
      file_base = "#{sample}.#{se}.#{temp[3]}_#{temp[5]}_#{temp[4]}"

      cmd = "ln -s #{fst} #{fastq_dir}/#{file_base}.fastq.gz"
      # print cmd
      `#{cmd}`

      if temp[4] == '1'
        fastq_se << "#{sample}.#{se}.#{temp[3]}_#{temp[5]}"
      end

    end
  end
  fastq_se = fastq_se.join(" ")

  ############### make run.sh in each sample dir
  if ! rna_flag
    thread = `gxpc e hostname | wc -l`.gsub(/\n/, '')
    `mkdir -p #{File.join( storage, run, sample)}`
    run_file = File.join( storage, run, sample, 'run.sh' )
    File.open(run_file, "w") do |f|
      f.write "gxpc make -k -j #{thread} -f #{path_makefile} "
      f.write "SAMPLE=#{sample} PE_READS=\"#{fastq_pe}\" SE_READS=\"#{fastq_se}\" "
      f.write "SURE_POS=#{sure_pos}"
      f.write " TARGET=genome" if genome_flag;
      f.write "\n";
    end

  end

end ########## END OF MAIN LOOP ; for sample in libids
end # end of function



########## make run script <auto_run_SUFFIX.sh>
def make_auto_run_sh_core( path_check, rna_flag)
  if ! rna_flag
    return <<EOS
\tcd $BASE_DIR/$DIR && date >> make.log && (time sh run.sh) >> make.log 2>> make.log
\tcd $BASE_DIR/ && ruby -W0 #{path_check} $DIR >> check_results.log 2>> check_results.log
EOS
  else <<EOS
\tqsub -sync y -N RNA-$DIR -o $DIR/make.log -j y /work/HiSeq2000/BaseCall/qsub_rna.sh $DIR
EOS
  end
end

def make_auto_run_sh(storage, slide,  group_by_prep,  path_check)

  run_script = File.join( storage, slide, "auto_run.#{Time.now().strftime("%Y%m%d-%H%M%S")}.sh" )
  File.open( run_script, 'w' ) do |f|
    f.write <<EOS
#!/usr/bin/env bash
BASE_DIR=#{File.join(storage,slide)}
EOS
    for regex, rows in group_by_prep
      suffix = $PREP.get_suffix( rows[0].prep_kit )
      rna_flag = (  /RNA/ =~ regex.to_s )? true : false
      # STDERR.puts '---rna_flag---', rna_flag
      f.write <<EOS
for DIR in #{ rows.map{|e| e.library_id + suffix }.join(" ") }
do
echo $DIR
EOS
      f.puts make_auto_run_sh_core( path_check, rna_flag )
      f.puts 'done'
    end
  end
  return run_script
end

